### PR TITLE
generic: 5.15: enable Werror by default for kernel compile

### PR DESCRIPTION
**Pull request by Ansuel**
From 5.15 and up linux kernel introduced CONFIG_WERROR to flag any warning as error. To improve code quality, enable this by default to catch any warning and fix it.

---

Very fun change... tought I enabled this when I pushed 5.15 but for some reason this got dropped later... Enable once and for all and see the magic CI explode.

### CODE REVIEW

**Comment by ptpt52** [comment]
mark.

**Comment by Ansuel** [comment]
Ohhh boy this will be EEEEEXTRA FUN (already had some result from my personal CI)

**Comment by john-tho** [comment]
Let us know if / how you want help here?

---

I saw the ipq40xx error: https://github.com/openwrt/openwrt/actions/runs/4899165678/jobs/8748928272
```
 make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-arm-openwrt-linux-muslgnueabi_musl/linux-ipq40xx_chromium/linux-5.15.110'
drivers/net/phy/phylink.c: In function 'phylink_get_linkmodes':
drivers/net/phy/phylink.c:360:9: error: enumeration value 'PHY_INTERFACE_MODE_PSGMII' not handled in switch [-Werror=switch]
  360 |         switch (interface) {
      |         ^~~~~~
cc1: all warnings being treated as errors
make[7]: *** [scripts/Makefile.build:289: drivers/net/phy/phylink.o] Error 1
```
Where it looks like our target [PSGMII patch](https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-5.15/704-net-phy-define-PSGMII-PHY-interface-mode.patch) will need to be updated for the [_phylink: add generic validate implementation_ `phylink_get_linkmodes`](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/backport-5.15/703-09-v5.17-net-phylink-add-generic-validate-implementation.patch) backport
@robimarko Does PSGMII do the same as QSGMII here, and support 1000, 100, and 10 at both half and full duplex? Thanks.

**Comment by neheb** [comment]
```
fs/jffs2/xattr.c: In function 'jffs2_build_xattr_subsystem':
fs/jffs2/xattr.c:887:1: error: the frame size of 1088 bytes is larger than 1024 bytes [-Werror=frame-larger-than=]
  887 | }
      | ^
```

hmmmm

**Comment by john-tho** [comment]
> ```
> fs/jffs2/xattr.c: In function 'jffs2_build_xattr_subsystem':
> fs/jffs2/xattr.c:887:1: error: the frame size of 1088 bytes is larger than 1024 bytes [-Werror=frame-larger-than=]
>   887 | }
>       | ^
> ```

There may be a patch for the [jffs2 xattrs stack warning](https://github.com/openwrt/openwrt/actions/runs/4899165678/jobs/8748931275?pr=12545) on lkml that linux-mtd has ignored since 2017…: https://lore.kernel.org/linux-mtd/20170509203003.11986-1-fabf@skynet.be/

**Comment by Ansuel** [comment]
> > ```
> > fs/jffs2/xattr.c: In function 'jffs2_build_xattr_subsystem':
> > fs/jffs2/xattr.c:887:1: error: the frame size of 1088 bytes is larger than 1024 bytes [-Werror=frame-larger-than=]
> >   887 | }
> >       | ^
> > ```
> 
> There may be a patch for the [jffs2 xattrs stack warning](https://github.com/openwrt/openwrt/actions/runs/4899165678/jobs/8748931275?pr=12545) on lkml that linux-mtd has ignored since 2017…: https://lore.kernel.org/linux-mtd/20170509203003.11986-1-fabf@skynet.be/

thanks for the link... I already prepared the same exact patch yesterday but better to have something already pushed... I reproposed the patch upstream... let see how it goes... if they are fast enough i won't have to put it in pending...

**Comment by robimarko** [comment]
`ipq40xx` is probably going to fail on sdhci-msm due to unused clock op as we patch the driver to use the generic op instead of the custom driver one as that one is failling on the XO clock somehow and I never managed to trace why its failling.

**Comment by Ansuel** [comment]
@robimarko TADA https://github.com/Ansuel/openwrt/actions/runs/4901610487/jobs/8752998402

**Comment by abajk** [comment]
[lantiq/ase fix](https://github.com/openwrt/openwrt/commit/146634fedea6c60c291ce0ece65af87ca73371f8)

**Comment by Ansuel** [comment]
@abajk can you reference the compilation warning? I'm following this pattern in almost all commit.

**Comment by abajk** [comment]
I see that there are still warnings from the pci driver on the lantiq/xrx200.

> @abajk can you reference the compilation warning? I'm following this pattern in almost all commit.

Should I create a separate PR, send the patch upstream, or do something other than that?

**Comment by Ansuel** [comment]
> I see that there are still warnings from the pci driver on the lantiq/xrx200.

i'm fixing each warning with a single commit. 

> Should I create a separate PR, send the patch upstream, or do something other than that?

I would first collect each required change and then see how to proceed with upstreaming.

**Comment by robimarko** [comment]
> @robimarko TADA https://github.com/Ansuel/openwrt/actions/runs/4901610487/jobs/8752998402

Well, I tried chasing the root cause 2 or 3 times before and it was always a dead end, everything pointed out at the XO clock but that is just a fixed clock and makes no sense.

**Comment by abajk** [comment]
Here are some [fixes](https://github.com/abajk/openwrt/commits/lantiq-ase-fix-warning) for lantiq. The patch for lzma-loader can probably be skipped.

**Comment by Ansuel** [comment]
@abajk nice!

We should check this better tho  https://github.com/abajk/openwrt/commit/b8d37b6309561789f31c7e059403b418107e96b2

**Comment by abajk** [comment]
> @abajk nice!
> 
> We should check this better tho [abajk@b8d37b6](https://github.com/abajk/openwrt/commit/b8d37b6309561789f31c7e059403b418107e96b2)

I think setting the mac address for the ethernet can be moved to [02_network](https://github.com/openwrt/openwrt/blob/master/target/linux/lantiq/xway/base-files/etc/board.d/02_network) and this part of the patch can be removed.

**Comment by abajk** [comment]
On the other hand, I don't have this device and I'm not sure if it will work the same way.

**Comment by Ansuel** [comment]
> On the other hand, I don't have this device and I'm not sure if it will work the same way.

i will pretend i didn't see the that patch honestly... notice way too much stuff done wrong... And that is outside the subject of this cleanup.

**Comment by Ansuel** [comment]
Almost there... only 3 target fail... (one hard to fix, 2 should be easy)

And then I have all the fun of testing each subtarget...

**Comment by abajk** [comment]
Do all kernel modules need to be verified too?

**Comment by Ansuel** [comment]
> Do all kernel modules need to be verified too?

CI is our friend and aside from some very specific one, every module is already OK...

**Comment by Ansuel** [comment]
YAY we are in subtarget wide testing phase... And already 2 fail....

**Comment by ptpt52** [comment]
You've worked so hard, that's awesome. Fixing these compilation warnings is a lot of work

**Comment by robimarko** [comment]
I am not sure when he sleeps

**Comment by Ansuel** [comment]
In theory now every subtarget should correctly compile with no error...

I also added to test each subtarget building everything so sorry for everyone pushing PR... the CI test will take probably days.... (but it's necessary)

**Comment by ptpt52** [comment]
github is now crash, somewhat causing by those actions aha!

Christian Marangi ***@***.***> 于 2023年5月9日周二 20:44写道：

**Comment by Ansuel** [comment]
> Some of these look like they should be fixed separately instead of adding an extra patch.

Mhhh some example? I'm trying to not touch the original patch where possibile (mainly the situation where i comment stuff)

**Comment by neheb** [comment]
Right. In the case of hack patches that we carry, makes sense to edit directly.

**Comment by Ansuel** [comment]
@neheb i'm in fixing spree so if you can help catch what i can directly fix!

**Comment by oliv3r** [comment]
@Ansuel I see you are touching a lot of the realtek target; which ironically some I have fixed on my local branch ages ago :)

I'll review those specific changes; but also makes it a little annoying as we know we have unused functions, as they serve as 'documentation' somehow, but they really should be removed/used.

What-a-mess :(

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c [hunk 1]
```diff
@@ -1427,7 +1428,7 @@ static int rtl83xx_fib_event(struct notifier_block *this, unsigned long event, v
 			fib_info_hold(fib_work->fen_info.fi);
 
 		} else if (info->family == AF_INET6) {
-			struct fib6_entry_notifier_info *fen6_info = ptr;
+			/*struct fib6_entry_notifier_info *fen6_info = ptr;*/
```
Comment (oliv3r) [inline target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c:1431]: we at least should be consistent in our commenting style; probably best to use 'illegal' slashes, as those will trigger other issues later (upstream rejecting it, grep ability for illegal comments etc).

**Comment by Ansuel** [comment]
@oliv3r this is really for the sake of enforcing WERROR with the next release and onwards... So I didn't drop them and just commented keeping the same format. Further modification would result in additional bug or unintended change

**Comment by oliv3r** [comment]
> @oliv3r this is really for the sake of enforcing WERROR with the next release and onwards... So I didn't drop them and just commented keeping the same format. Further modification would result in additional bug or unintended change

Oh I get it, and it is a nice thing :) I had fixed most warnings (not unused thingers) half a year ago, they are in my queue :package:

**Comment by abajk** [comment]
CI skips the UML target. It is not marked as source-only. Is this intentional?

File: target/linux/ramips/patches-5.15/835-asoc-add-mt7620-support.patch [hunk 2]
```diff
@@ -945,7 +945,11 @@ Signed-off-by: John Crispin <blogic@openwrt.org>
 +
 +	ralink_i2s_init_dma_data(i2s, res);
 +
-+	device_reset(&pdev->dev);
++	ret = device_reset(&pdev->dev);
++	if (ret) {
++		dev_err(&pdev->dev, "cfailed to reset device\n");
```
Comment (ThiloteE) [inline target/linux/ramips/patches-5.15/835-asoc-add-mt7620-support.patch:950]: typo? - "cfailed" to "failed"

**Comment by Ansuel** [comment]
I may or may not have finished fixing everything broken... Fun thing I got totally sidetracked with fixing the broadcom-wl driver as i totally didn't notice it was broken from the start from ages...

**Comment by ptpt52** [comment]
@Ansuel 
build failed for `xtables-addons` since this PR merged.


```
/mnt/Data/Sources/openwrt/x-wrt/build_dir/target-aarch64_cortex-a53_musl/linux-mediatek_filogic/xtables-addons-3.21/extensions/LUA/byte_array.c:112:9: error: ISO C90 forbids variable length array 'res' [-Werror=vla]
  112 |         char res[255 + (array->length * 3)]; /* make sure the buffer is big enough*/
      |         ^~~~
```

**Pull request merged**